src/cid.rs

```diff
@@ -91,34 +91,57 @@ where
     }
 
     fn to_string_v1(&self) -> String {
-        multibase::encode(Base::Base32Lower, self.to_bytes().as_slice())
+        multibase::encode(Base::Base32Lower, &self.to_bytes())
     }
 
-    fn to_bytes_v0(&self) -> Vec<u8> {
-        self.hash.to_vec()
+    fn copy_to_slice_v0(&self, buffer: &mut [u8]) -> usize {
+        let hash_end = self.hash.len();
+        (&mut buffer[..hash_end]).copy_from_slice(&self.hash);
+        hash_end
     }
 
-    fn to_bytes_v1(&self) -> Vec<u8> {
-        let mut res = Vec::with_capacity(16);
-
+    fn copy_to_slice_v1(&self, buffer: &mut [u8]) -> usize {
         let mut buf = varint_encode::u64_buffer();
         let version = varint_encode::u64(self.version.into(), &mut buf);
-        res.extend_from_slice(version);
+        let version_end = version.len();
+
         let mut buf = varint_encode::u64_buffer();
         let codec = varint_encode::u64(self.codec.into(), &mut buf);
-        res.extend_from_slice(codec);
-        res.extend_from_slice(&self.hash);
+        let codec_end = version_end + codec.len();
+
+        let hash_end = codec_end + self.hash.len();
 
-        res
+        (&mut buffer[..version_end]).copy_from_slice(&version);
+        (&mut buffer[version_end..codec_end]).copy_from_slice(&codec);
+        (&mut buffer[codec_end..hash_end]).copy_from_slice(&self.hash);
+
+        hash_end
     }
 
-    /// Convert CID to encoded bytes.
-    pub fn to_bytes(&self) -> Vec<u8> {
+    /// Copy encoded bytes into buffer.
+    ///
+    /// Will panic if the buffer is too small.
+    pub fn copy_to_slice(&self, buffer: &mut [u8]) -> usize {
         match self.version {
-            Version::V0 => self.to_bytes_v0(),
-            Version::V1 => self.to_bytes_v1(),
+            Version::V0 => self.copy_to_slice_v0(buffer),
+            Version::V1 => self.copy_to_slice_v1(buffer),
         }
     }
+
+    /// Max required buffer length.
+    pub fn max_buffer_len(&self) -> usize {
```

Comment:
Buffer seems like the wrong terminology here. I think something like `may_len` would make more sense.